src/bundle-manifest.ts

```diff
@@ -0,0 +1,160 @@
+ *
+ * Would produce an array of three bundles:
+ *   `[a]->[b,d], [e]->[f], [a,e]->[c]`
+ */
+export function generateBundles(depsIndex: IFileTransDepsIndex): IBundle[] {
+  const bundles: IBundle[] = [];
+  const invertedIndex = invertFileTransDepsIndex(depsIndex);
+
+  for (let [dep, entrypoints] of invertedIndex.entries()) {
+    const entrypointsArray = Array.from(entrypoints);
+    let bundle = bundles.find((bundle) => {
+      return bundle.entrypoints.size === entrypoints.size &&
+          entrypointsArray.every((e) => bundle.entrypoints.has(e));
+    });
+    if (!bundle) {
+      bundle = new Bundle(entrypoints);
+      bundles.push(bundle);
+    }
+    bundle.files.add(dep);
+  }
+  return bundles;
+}
+
+/**
+ * Generates a strategy function to merge all bundles where the dependencies
+ * for a bundle are shared by at least 2 entrypoints (default; set
+ * `minEntrypoints` to change threshold).
+ *
+ * This function will convert an array of 4 bundles:
+ *   `[a]->[b], [a,c]->[d], [c]->[e], [f,g]->[h]`
+ *
+ * Into the following 3 bundles, including a single bundle for all of the
+ * dependencies which are shared by at least 2 entrypoints:
+ *   `[a]->[b], [c]->[e], [a,c,f,g]->[d,h]`
+ */
+export function generateSharedDepsMergeStrategy(minEntrypoints: number):
+    IBundleStrategy {
+  return (bundles: IBundle[]): IBundle[] => {
+    const newBundles: IBundle[] = [];
+    let sharedBundle: Bundle|undefined;
+    bundles.forEach((bundle) => {
+      if (bundle.entrypoints.size >= minEntrypoints) {
+        sharedBundle =
+            sharedBundle ? mergeBundles([sharedBundle, bundle]) : bundle;
+      } else {
+        const newBundle = new Bundle();
+        bundle.entrypoints.forEach((e) => newBundle.entrypoints.add(e));
+        bundle.files.forEach((f) => newBundle.files.add(f));
+        newBundles.push(newBundle);
+      }
+    });
+    if (sharedBundle) {
+      newBundles.push(sharedBundle);
+    }
+    return newBundles;
+  };
+}
+
+/**
+ * A bundle strategy function which merges all shared dependencies into a
+ * bundle for an application shell.
+ */
+export function generateShellMergeStrategy(shell: URL): IBundleStrategy {
+  return (bundles: IBundle[]): IBundle[] => {
+    const newBundles = singleSharedDepsStrategy(bundles);
+    const shellBundle = newBundles.find((bundle) => bundle.files.has(shell));
+    const sharedBundle =
+        newBundles.find((bundle) => bundle.entrypoints.size > 1);
+    if (shellBundle && sharedBundle && shellBundle !== sharedBundle) {
+      newBundles.splice(newBundles.indexOf(shellBundle), 1);
+      newBundles.splice(newBundles.indexOf(sharedBundle), 1);
+      newBundles.push(mergeBundles([shellBundle, sharedBundle]));
+    }
+    return newBundles;
+  };
+}
+
+/**
+ * Returns an inverted FileTransDepsIndex where the keys are the dependencies
+ * and the values are the set of files which are dependent on them.
+ * `{a:[b,c], d:[b,e]}` becomes `{a:[a], b:[a,b,d], c:[a], d:[d], e:[d]}`.
```

Comment:
I was trying to avoid explicit identity dependency in the initial TransitiveDependenciesMap but I don't have a good reason to keep trying to do that.  Things are cleaner if they're included.  Updating code and comments to that end.